### PR TITLE
Enable TCP route registration

### PR DESCRIPTION
**Pull request by iredelmeier**
Thanks for contributing to 'routing-release'. To speed up the process of reviewing your pull request please provide us with:

* A short explanation of the proposed change:

In conjunction with https://github.com/cloudfoundry/route-registrar/pull/23, this change allows TCP routes to be registered through the route-registrar.

* An explanation of the use cases your change solves

Routes for CF components and other Bosh releases may need to be registered directly with the TCP router, rather than through CAPI, if their lifecycle does not align properly with CAPI's. This change allows these routes to be easily registered through the route-registrar.

* Instructions to functionally test the behavior change using operator interfaces (BOSH manifest, logs, curl, and metrics)

Assuming that the TCP router and route-registrar are configured per `cf-deployment`, you can register a new TCP route as with [this Perm ops file](https://github.com/cloudfoundry/cf-deployment/blob/027ae286ff9720226b6f5fa101d43cb5c6322e27/operations/experimental/perm-service-with-tcp.yml).

* Expected result after the change

* Current result before the change

* Links to any other associated PRs

https://github.com/cloudfoundry/route-registrar/pull/23

* [x] I have viewed signed and have submitted the Contributor License Agreement

* [x] I have made this pull request to the `develop` branch

* [x] I have run all the unit tests using `scripts/run-unit-tests`
  * (failed for unrelated reasons)

* [x] I have run Routing Acceptance Tests and Routing Smoke Tests on bosh lite

* [x] I have run CF Acceptance Tests on bosh lite

### CODE REVIEW

File: jobs/route_registrar/spec [hunk 1]
```diff
@@ -36,31 +37,52 @@ properties:
...
+    description: (optional, string) The OAuth server's URL. This is required to register any TCP routes.
+    default: https://uaa.service.cf.internal:8443
+  route_registrar.routing_api.client_id:
+    description: (optional, string) An OAuth client ID for a client that is permitted to add new TCP routes. This is required to register any TCP routes.
+    default: routing_api_client
+  route_registrar.routing_api.client_secret:
+    description: (optional, string) The OAuth client secret for the above client. This is required to register any TCP routes.
+  route_registrar.routing_api.ca_certs:
+    description: (optional, array of strings) The certificate authority certificates for any APIs that the route registrar is communicating with over HTTPS, e.g., the OAuth server. This is required to register any TCP routes.
+  route_registrar.routing_api.skip_cert_validation:
```
Comment (shalako) [inline jobs/route_registrar/spec]: I don't think this is necessary. Notice that other jobs in routing-release respect a top-level, job agnostic property `skip_ssl_validation`

File: jobs/route_registrar/spec [hunk 2]
```diff
@@ -36,31 +37,52 @@ properties:
...
-        health_check (optional, object): Script executed on frequency of `registration_interval`. If
+        health_check (optional, object, for all routes): Script executed on frequency of `registration_interval`. If
           healthcheck script exits with success, route registration heartbeat is sent. If script exits
-          with error, the route is unregistered. 
+          with error, the route is unregistered.
+        router_group (required, string, for tcp routes): Name of the router group to which the TCP route should be added.
+        backend_host (required, string, for tcp routes): DNS or IP address of the backend server that the TCP route should be routed to.
+        backend_port (required, string, for tcp routes): Port of the backend server that the TCP route should be routed to.
```
Comment (shalako) [inline jobs/route_registrar/spec]: I believe this is unnecessary. The `port` field above specifies the port on the backend that the router will forward TCP connections to.

File: jobs/route_registrar/spec [hunk 3]
```diff
@@ -36,31 +37,52 @@ properties:
...
           prepended with the instance index. e.g. 'some-uri.system-domain.com' will become '0-some-uri.system-domain.com' on the instance with index 0, and '2-some-url.system-domain.com' on the instance with index 2. When this value is enabled, each instance will register its own, unique, set of uris. To additionally continue to register these original uris, create another route with the same uris and set 'prepend_instance_index' to false (or omit the key entirely).
-        health_check (optional, object): Script executed on frequency of `registration_interval`. If
+        health_check (optional, object, for all routes): Script executed on frequency of `registration_interval`. If
           healthcheck script exits with success, route registration heartbeat is sent. If script exits
-          with error, the route is unregistered. 
+          with error, the route is unregistered.
+        router_group (required, string, for tcp routes): Name of the router group to which the TCP route should be added.
+        backend_host (required, string, for tcp routes): DNS or IP address of the backend server that the TCP route should be routed to.
```
Comment (shalako) [inline jobs/route_registrar/spec]: Do you need to register a TCP route for a host other than the one on which route_registrar runs? If not, this is not necessary; route_registrar automatically detects the IP of the VM.

Comment (shalako) [reply jobs/route_registrar/spec]: If you do need to specify the backend host, you can leverage the existing top-level `host` property: https://github.com/cloudfoundry/routing-release/blob/develop/jobs/route_registrar/spec#L35-L37

File: jobs/route_registrar/spec (hunk 2 above)
Comment (jeffpak) [reply jobs/route_registrar/spec]: I'm not sure I follow. Did you want a user to specify another route solely for a separate backend port? Per [this comment on the previous PR ](https://github.com/cloudfoundry/route-registrar/pull/21#issuecomment-349850970), it seemed more convenient to just specify an external port and a backend port.

**Comment by jeffpak** [comment]
Hello @shalako, we recently updated our PR to reflect the changes you've requested on both route registrar and the routing release. If you could take a look over them again and let us know of any other concerns, that would be swell!

Regards,
Perm Team

**Pull request closed without merging**